applications/sdp/mspi/src/main.c

```diff
@@ -52,215 +57,494 @@ typedef struct __packed {
 	case MSPI_CPP_MODE_0: {
 		vio_config.clk_polarity = 0;
-		nrf_vpr_csr_vio_out_set(PIN_OUT_LOW_MASK(VIO(NRFE_MSPI_SCK_PIN_NUMBER)));
+		out = BIT_SET_VALUE(out, pin_to_vio_map[NRFE_MSPI_SCK_PIN_NUMBER],
+				    VPRCSR_NORDIC_OUT_LOW);
+		xfer_ll_params.eliminate_last_pulse = false;
 		break;
 	}
 	case MSPI_CPP_MODE_1: {
 		vio_config.clk_polarity = 1;
-		nrf_vpr_csr_vio_out_set(PIN_OUT_LOW_MASK(VIO(NRFE_MSPI_SCK_PIN_NUMBER)));
+		out = BIT_SET_VALUE(out, pin_to_vio_map[NRFE_MSPI_SCK_PIN_NUMBER],
+				    VPRCSR_NORDIC_OUT_LOW);
+		xfer_ll_params.eliminate_last_pulse = true;
 		break;
 	}
 	case MSPI_CPP_MODE_2: {
 		vio_config.clk_polarity = 1;
-		nrf_vpr_csr_vio_out_set(PIN_OUT_HIGH_MASK(VIO(NRFE_MSPI_SCK_PIN_NUMBER)));
+		out = BIT_SET_VALUE(out, pin_to_vio_map[NRFE_MSPI_SCK_PIN_NUMBER],
+				    VPRCSR_NORDIC_OUT_HIGH);
+		xfer_ll_params.eliminate_last_pulse = false;
 		break;
 	}
 	case MSPI_CPP_MODE_3: {
 		vio_config.clk_polarity = 0;
-		nrf_vpr_csr_vio_out_set(PIN_OUT_HIGH_MASK(VIO(NRFE_MSPI_SCK_PIN_NUMBER)));
+		out = BIT_SET_VALUE(out, pin_to_vio_map[NRFE_MSPI_SCK_PIN_NUMBER],
+				    VPRCSR_NORDIC_OUT_HIGH);
+		xfer_ll_params.eliminate_last_pulse = true;
 		break;
 	}
 	}
-
+	nrf_vpr_csr_vio_out_set(out);
 	nrf_vpr_csr_vio_config_set(&vio_config);
 }
 
-void prepare_and_send_data(uint8_t *data, uint8_t data_length)
+static void prepare_and_send_data()
 {
-	memcpy(&(data_buffer[2]), data, data_length);
+	NRFX_ASSERT(nrfe_mspi_dev_cfg.ce_num < ce_vios_count);
+
+	struct xfer_io_mode_cfg xfer_modes = get_io_modes(nrfe_mspi_dev_cfg.io_mode);
+	uint32_t data;
 
 	/* Send command */
-	xfer_ll_params.ce_hold = true;
-	xfer_ll_params.word_size = mspi_dev_configs.cmd_length;
-	xfer_ll_params.data_len = 1;
-	xfer_ll_params.data_to_send = data_buffer;
 
-	if (mspi_dev_configs.io_mode == MSPI_IO_MODE_QUAD) {
-		nrf_vpr_clic_int_pending_set(NRF_VPRCLIC, VEVIF_IRQN(HRT_VEVIF_IDX_WRITE_QUAD));
-	} else {
-		nrf_vpr_clic_int_pending_set(NRF_VPRCLIC, VEVIF_IRQN(HRT_VEVIF_IDX_WRITE_SINGLE));
-	}
+	/* TODO: Jira ticket: NRFX-6703
+	 * Device waits this time after setting CE and before sending
+	 * first bit, make this value dependent on dummy cycles.
+	 */
+	xfer_ll_params.counter_initial_value = 120;
+	data = nrfe_mspi_xfer_packet.cmd << (BITS_IN_WORD - nrfe_mspi_dev_cfg.cmd_length);
+	xfer_ll_params.data = (uint8_t *)&data;
+	xfer_ll_params.ce_vio = ce_vios[nrfe_mspi_dev_cfg.ce_num];
+	xfer_ll_params.ce_hold = nrfe_mspi_xfer.hold_ce;
+	xfer_ll_params.ce_polarity = nrfe_mspi_dev_cfg.ce_polarity;
+	xfer_ll_params.bit_order = HRT_BO_REVERSED_WORD;
+
+	ll_prepare_transfer(xfer_modes.command, nrfe_mspi_dev_cfg.cmd_length);
+
+	nrf_vpr_clic_int_pending_set(NRF_VPRCLIC, VEVIF_IRQN(HRT_VEVIF_IDX_WRITE));
 
 	/* Send address */
-	xfer_ll_params.word_size = mspi_dev_configs.addr_length;
-	xfer_ll_params.data_to_send = data_buffer + XFER_ADDRESS_IDX;
+	data = nrfe_mspi_xfer_packet.address << (BITS_IN_WORD - nrfe_mspi_dev_cfg.addr_length);
```

Comment:
The `cmd_length` and `addr_length` should be in bytes, not bits!

Reply:
@jaz1-nordic is that specified somewhere in Zephyr documentation? I am asking, because I did not find, if it's supposed to be in bits or bytes when I was creating my draft, so I assumed it depends on driver implementation and decided it should be in bits to include the one bit before command in 9-bit mode for the display that we own (see pictures that I sent on the chat of the last meeting. However, I misunderstood the datasheet: I thought that only the first byte has one additional bit at the beginning, but it turns out that all data is sent in 9-bit words in this mode and the first-bit value should probably be determined by the user of MSPI (it is a control signal). So, I guess we can assume, for now, that data, address, and command have all sizes specified in bytes and deal with the 9-bit issue on later notice. This would also resolve the problem with sending data on only one clock cycle, as in quad mode there will always be at least two cycles needed. It will only be still a problem for octal mode, but firstly, octal can be fixed later, as we are focusing on quad right now, and secondly, I doubt there could be a transmission of only one byte, at least it's unlikely, in my opinion.
Unless it is specified in the documentation, that all sizes should be in bytes, then this definitely needs to be changed.

Reply:
I agree, there needs to be a way to send lengths different than multiples of 8 bits, otherwise communication with some devices will be impossible. I have prepared low level and HRT to be able to handle 1 clock pulse transfers and they count everything in bits.

So lets assume for now that byte resolution is ok and when higher resolution will be needed it will be quite easy to fix.

Reply:
changed to bytes